File: episodes/2023/002/win32circle.go
package main

import (
	"fmt"
	"syscall"
	"unsafe"

	"github.com/lxn/win"
)

func main() {
	var hwnd win.HWND
	var hdc win.HDC
	var hbrush win.HBRUSH

	// Register the window class
	var wc win.WNDCLASSEX
	wc.CbSize = uint32(unsafe.Sizeof(wc))
	wc.LpfnWndProc = syscall.NewCallback(wndProc)
	wc.HInstance = win.HINSTANCE(syscall.Handle(win.GetModuleHandle(nil)))
	wc.HIcon = win.LoadIcon(0, win.MakeIntResource("IDI_APPLICATION"))
	wc.HCursor = win.LoadCursor(0, win.MakeIntResource("IDC_ARROW"))
	wc.HbrBackground = win.HBRUSH(win.GetStockObject(win.WHITE_BRUSH))
	wc.LpszClassName = syscall.StringToUTF16Ptr("Circle")
	win.RegisterClassEx(&wc)

	// Create the window
	hwnd = win.CreateWindowEx(0, syscall.StringToUTF16Ptr("Circle"), syscall.StringToUTF16Ptr("Circle"), win.WS_OVERLAPPEDWINDOW, win.CW_USEDEFAULT, win.CW_USEDEFAULT, win.CW_USEDEFAULT, win.CW_USEDEFAULT, 0, 0, wc.HInstance, nil)
	win.ShowWindow(hwnd, win.SW_SHOWNORMAL)

	// Run the message loop
	var msg win.MSG
	for win.GetMessage(&msg, 0, 0, 0) != 0 {
		win.TranslateMessage(&msg)
		win.DispatchMessage(&msg)
	}
}

func wndProc(hwnd win.HWND, msg uint32, wparam, lparam uintptr) uintptr {
	switch msg {
	case win.WM_PAINT:
		var ps win.PAINTSTRUCT
		hdc := win.BeginPaint(hwnd

File: episodes/2023/002/win32circle.c
// cl win32circle.c User32.lib gdi32.lib

#include <windows.h>

LRESULT CALLBACK WindowProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
{
    switch (uMsg)
    {
    case WM_PAINT:
    {
        PAINTSTRUCT ps;
        HDC hdc = BeginPaint(hwnd, &ps);

        // Set the color of the circle to red
        HBRUSH hbrush = CreateSolidBrush(RGB(255, 0, 0));
        SelectObject(hdc, hbrush);

        // Draw a circle with a radius of 50 pixels
        Ellipse(hdc, 50, 50, 100, 100);

        // Clean up
        DeleteObject(hbrush);
        EndPaint(hwnd, &ps);
        return 0;
    }
    case WM_DESTROY:
    {
        PostQuitMessage(0);
        return 0;
    }
    }

    return DefWindowProc(hwnd, uMsg, wParam, lParam);
}

int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nCmdShow)
{
    WNDCLASSEX wc = {sizeof(WNDCLASSEX), CS_HREDRAW | CS_VREDRAW, WindowProc, 0, 0, hInstance, LoadIcon(NULL, IDI_APPLICATION), LoadCursor(NULL, IDC_ARROW), (HBRUSH)(COLOR_WINDOW + 1), NULL, "Circle", LoadIcon(NULL, IDI_APPLICATION)};
    RegisterClassEx(&wc);

    HWND hwnd = CreateWindowEx(0, "Circle", "Circle", WS_OVERLAPPEDWINDOW, CW_USEDEFAULT, CW_USEDEFAULT, CW_USEDEFAULT, CW_USEDEFAULT, NULL, NULL, hInstance, NULL);
    ShowWindow(hwnd, nCmdShow);

    MSG msg;
    while (GetMessage(&msg, NULL, 0, 0))
    {
        TranslateMessage(&msg);
        DispatchMessage(&msg);
    }

    return (int)msg.wParam;
}
